### PR TITLE
[backend] Add garbage collection manager (#1536)

### PR DESCRIPTION
**Pull request by JeremyCloarec**
### Proposed changes

* A garbage collection manager has been added to permanently delete deleteOperation older than the configured retention time

### Related issues

* part of #1536 

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case (coverage and e2e)
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 1]
```diff
@@ -3405,7 +3405,8 @@ export const internalDeleteElementById = async (context, user, id, opts = {}) =>
       // Start by deleting external files
       const isTrashableElement = !isInferredIndex(element._index)
                 && (isStixCoreObject(element.entity_type) || isStixCoreRelationship(element.entity_type) || isStixSightingRelationship(element.entity_type));
-      const forceDelete = !!opts.forceDelete || !isTrashableElement;
+      // isModuleActivated from domain/settings cannot be used because of dependency cycle issues, we use the flag from conf instead
+      const forceDelete = !!opts.forceDelete || !isTrashableElement || !ENABLED_GARBAGE_COLLECTION_MANAGER;
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/database/middleware.js]: this part is not right, ENABLED_GARBAGE_COLLECTION_MANAGER can be false in a node but true in another (cluster mode). 
And we might need to validate this condition also (are we sure we want to delete permanently when garbage collector is disabled?)

File: opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  const context = executionContext('garbage_collection_manager');
+  const deletedRentionDaysToConsider = Math.max(DELETED_RETENTION_DAYS, 1);
+  const deleteOperationsToManage = await findOldDeleteOperations(context, GARBAGE_COLLECTION_MANAGER_USER, deletedRentionDaysToConsider, BATCH_SIZE);
+  let errorCount = 0;
+  for (let i = 0; i < deleteOperationsToManage.length; i += 1) {
+    try {
+      const deleteOperation = deleteOperationsToManage[i];
+      await completeDelete(context, GARBAGE_COLLECTION_MANAGER_USER, deleteOperation.id);
+    } catch (e) {
+      logApp.warn(e, `[OPENCTI-MODULE] Error when processing delete operation for ${deleteOperationsToManage[i].id}, skipping.`);
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts]: logger must never build dynamic message, meta info need to be used.
logApp.warn(message, { id: deleteOperationsToManage[i].id })

File: opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ */
+export const garbageCollectionHandler = async () => {
+  const context = executionContext('garbage_collection_manager');
+  const deletedRentionDaysToConsider = Math.max(DELETED_RETENTION_DAYS, 1);
+  const deleteOperationsToManage = await findOldDeleteOperations(context, GARBAGE_COLLECTION_MANAGER_USER, deletedRentionDaysToConsider, BATCH_SIZE);
+  let errorCount = 0;
+  for (let i = 0; i < deleteOperationsToManage.length; i += 1) {
+    try {
+      const deleteOperation = deleteOperationsToManage[i];
+      await completeDelete(context, GARBAGE_COLLECTION_MANAGER_USER, deleteOperation.id);
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts:24]: In completeDelete you manage  await deleteAllObjectFiles.
Curious how do you manage files part of "file indexing feature".
What occurs when the user put the entity in the bin? Data indexed from files inside this entity are still available for search?

File: opencti-platform/opencti-graphql/src/database/middleware.js (hunk 1 above)
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/database/middleware.js]: I agree, this is actually a functional question.

File: opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts (hunk 3 above)
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts:24]: We introduced a flag "removed" turned to true when item goes in the bin. "Removed" files are excluded from search.
We turn off the flag on restore or delete properly the file on permanent deletion.

Comment (Kedae) [reply opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts:24]: The data already indexed from an entity in the bin is marked as "unavailable" and doesn't come out on search. When the entity is deleted forever, the indexation is cleaned.

Comment (Kedae) [reply opencti-platform/opencti-graphql/src/manager/garbageCollectionManager.ts:24]: Speaking under the team's control, but that's what I understood :)

**Review by labo-flg — APPROVED** [review]
tested ok

**Review by SouadHadjiat — APPROVED** [review]

**Pull request merged**